src/Orleans.Core/Utils/KeyedService.cs

```diff
@@ -6,26 +6,53 @@
 
 namespace Orleans.Runtime
 {
-    public interface IKeyedService<TKey, out TService> : IEquatable<TKey>
-    {
-        TService GetService(IServiceProvider services);
-    }
-
     public class KeyedService<TKey, TService, TInstance> : IKeyedService<TKey, TService>
             where TInstance : TService
     {
-        private readonly TKey key;
+        public TKey Key { get; }
 
         public KeyedService(TKey key)
         {
-            this.key = key;
+            this.Key = key;
         }
 
         public TService GetService(IServiceProvider services) => services.GetService<TInstance>();
 
         public bool Equals(TKey other)
         {
-            return Equals(key, other);
+            return Equals(Key, other);
+        }
+    }
+
+    public class KeyedSingletonService<TKey, TService> : IKeyedService<TKey, TService>
+    where TService : class
+    {
+        Func<IServiceProvider, TKey, TService> factory;
+        private TService instance;
```

Comment:
Is there scope for concurrency issues? Maybe we need a lock in `GetService`?

Reply:
Will look into.